text/0000-mutex_traits.md

```diff
@@ -0,0 +1,174 @@
+|               | where T: ?Sized   | by default for Sized |                    |
+
+## Defining ?Trait and !Trait
+
+`?Trait` and `!Trait` act as if they were marker traits. They define no methods and have no associated items. They are defined implicitly in the same scope as the definition of `Trait` and imported wherever `Trait` is imported.
+
+## Implementing ?Trait
+
+`?Trait` can only be implemented for types for which a default impl has been explicitly defined (e.g. `Send` and `Sync`). Explicit default impls of `?Trait` are not allowed. `?Trait` is implemented by default anyway, and it would not make sense to implement it except in the cases where an explicit default impl exists. As a rule, the syntax `?Trait` will be very uncommon.
+
+`?Trait` follows the same orphan rules as `Trait`.
+
+## Implementing !Trait
+
+Implementing `!Trait` for `T` is a forward-compatible guarantee that `T` does not and will not implement `Trait`. This makes negative reasoning explicit and avoids backwards compatibility hazards. It goes without saying that it would be a coherence violation for a single type to implement both `Trait` and `!Trait`.
+
+`!Trait` follows the same orphan rules as `Trait`.
+
+## Bounding by !Trait
+
+Bounding by `!Trait` requires that types _explicitly implement_ `!Trait` in order to meet that bound. As mentioned prior, this avoids the hazard that implicit negative reasoning introduces.
+
+## Syntactic sugar: Implicit `!Trait` inference
+
+If a type `T` implements a trait `Foo` which is bounded `!Bar`, an implementation of `!Bar` is inferred for `T` (unless `T` explicitly implements `!Bar`, of course). This avoids boilerplate negative impls which are inferrable from other impls for the type.
+
+## Clarification of default impl rules
+
+If a default impl of `Trait` exists, these rules are used to determine the relation between `T` and `Trait`:
+
+* If `Trait`, `?Trait` or `!Trait` is implemented for `T`, that impl defines the relation
+* If one of the members of `T` impls a trait which conflicts with the default impl, `T` is `?Trait`
+* Otherwise, `T` implements the default impl.
+
+Note that this set of rules is sound if we suppose that every trait has an implicit default impl of `?Trait`.
+
+## Orphan rule warbles
+
+The rules above all apply to a Rust system as a whole, composed of multiple crates associated as a directed acyclic graph. Within crates and modules, orphan rules allow silence to have a semantic expressions that is slightly different from these rules. Unfortunately, eliminating this warble would be backwards incompatible.
+
+Specifically, when both a trait and a type are defined within a single crate, that type and trait have the relationship `T: !Trait` by default, rather than `?Trait`, only within that crate. This allows a certain degree of implicit negative reasoning which cannot be performed outside of that local context. It does not present a logical contradiction for this proposal.
```

Comment:
I'd like to have "unless T actually has an `impl` for `Trait` in addition to 'by default', to clarify the meaning.